### PR TITLE
Update form name by storing name on FormDef instead of Form

### LINKED ISSUE #355: When there is a published form version, don't update form name from new drafts
https://github.com/getodk/central-backend/pull/344 updates the form name based on new form definitions. It does this update every time a new draft is created. This becomes confusing if the title change is anything other than a typo fix and especially confusing if a draft with a title change is created and then abandoned.

The expected behavior is that the title of the most recent draft is used until there's a published version at which point the title of the most recent published version is always used whether or not there's an active draft.

### PR DESCRIPTION
**Pull request by ktuite**
Closes #355 by moving a form's `name` (or <title> as defined in the form XML) from Form to FormDef. Form will always reference the appropriate FormDef and get the name from there.

This PR includes a migration that adds `name` to the `form_defs` table and backfills it with form title information read from form XML. It also includes a migration to drop the `name` field from the `forms` table. (Er, I was nervous about dropping that field so I just changed the column name but will change the migration to drop the column before I finalize this PR.)

One thing that changes is that a little bit more information about an unpublished form comes through the API (like its `version` and `sha`) when that information used to be `null`. It is still possible to determine if a form is unpublished by looking at its `publishedAt` field, though. We should probably add a note about this to the docs.

Endpoints that only show published forms (like the OpenRosa `formList`) are unaffected and work the same as before.

### CODE REVIEW

File: lib/model/migrations/20210423-02-drop-form-name.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+// except according to the terms contained in the LICENSE file.
+
+const { Form } = require('../frames');
+
+const up = async (db) => {
+  await db.raw('ALTER TABLE forms RENAME COLUMN name to nonactive_name');
+};
+
+const down = async (db) => {
+  await db.raw('ALTER TABLE forms RENAME COLUMN nonactive_name to name');
```
Comment (ktuite) [inline lib/model/migrations/20210423-02-drop-form-name.js:17]: I need to change this migration to just drop the column

**Comment by ktuite** [comment]
The frontend COULD show the historical form names, now, and not just the version...

Screen Shot 2021-04-23 at 12 31 07 PM

File: lib/model/frames/form.js [hunk 2]
```diff
@@ -112,7 +113,7 @@ class Form extends Frame.define(
       const version = versionText.orElse('');
       const name = nameText.orNull();
       const key = pubKey.map((k) => new Key({ public: k }));
-      return new Form.Partial({ xmlFormId, name }, { def: new Form.Def({ version }), key });
+      return new Form.Partial({ xmlFormId, name }, { def: new Form.Def({ version, name: name }), key });
```
Comment (issa-tseng) [inline lib/model/frames/form.js:116]: probably if you `make lint` you'll get a complaint about this line.

File: lib/model/frames/form.js [hunk 3]
```diff
@@ -50,10 +51,10 @@ class Form extends Frame.define(
   table('forms'),
   'id',                                 'projectId',    readable,
   'xmlFormId',    readable,             'state',        readable, writable,
-  'name',         readable,             'currentDefId',
```
Comment (issa-tseng) [inline lib/model/frames/form.js:53]: i'd probably rather just name get taken out here and the rest left alone to preserve the line history.

File: lib/model/query/forms.js [hunk 4]
```diff
@@ -241,8 +242,9 @@ order by coalesce(forms.name, forms."xmlFormId") asc`)
 /* eslint-disable indent */
 const versionJoinCondition = (version) => (
   (version === '___') ? versionJoinCondition('') :
-  (version == null) ? sql`form_defs.id=forms."currentDefId"` :
+  (version == null) ? sql`form_defs.id=coalesce(forms."currentDefId", forms."draftDefId")` :
```
Comment (issa-tseng) [inline lib/model/query/forms.js:245]: ah brilliant

File: lib/model/frames/form.js [hunk 5]
```diff
@@ -140,13 +142,14 @@ Form.Partial = class extends Form {};
...
-  'excelContentType',   readable
+  'excelContentType',   readable,       'name',           readable
 ) {
   forApi() {
     return {
       submissions: this.submissions || 0,
       lastSubmission: this.lastSubmission,
-      excelContentType: this.excelContentType
+      excelContentType: this.excelContentType,
+      name: this.name,
```
Comment (issa-tseng) [inline lib/model/frames/form.js:152]: why is this necessary? does it not come through the def in some cases?

File: lib/model/query/forms.js [hunk 6]
```diff
@@ -106,9 +106,10 @@ const createVersion = (partial, form, publish = false) => ({ run, one, FormAttac
         ? def.with({ publishedAt: new Date(), xml: partial.xml })
         : def.with({ draftToken: _getDraftToken(form), xml: partial.xml })))
       .then(compose(one, insert))
-      .then(ignoringResult((savedDef) => ((publish === true)
-        ? Forms._update(form, { currentDefId: savedDef.id, name: partial.name })
-        : Forms._update(form, { draftDefId: savedDef.id, name: partial.name })))),
+      .then(ignoringResult((savedDef) =>
+        ((publish === true)
+          ? Forms._update(form, { currentDefId: savedDef.id })
+          : Forms._update(form, { draftDefId: savedDef.id })))),
```
Comment (issa-tseng) [inline lib/model/query/forms.js:112]: love this diff

Comment (issa-tseng) [reply lib/model/query/forms.js:112]: well, i'm not sure why the whitespace change but i love the rm bits

**Review by issa-tseng — CHANGES_REQUESTED** [review]
just a couple questions then i'll do a squash merge

File: lib/model/frames/form.js (hunk 3 above)
Comment (ktuite) [reply lib/model/frames/form.js:53]: done!

File: lib/model/query/forms.js (hunk 6 above)
Comment (ktuite) [reply lib/model/query/forms.js]: oops, all fixed up now

File: lib/model/frames/form.js (hunk 5 above)
Comment (ktuite) [reply lib/model/frames/form.js]: you're right, this isn't necessary... something i tried along the way, i think

**Comment by issa-tseng** [comment]
merged via https://github.com/getodk/central-backend/commit/ee7ba91488e6b3d83e2d0d69eed09e6143902e90
